### PR TITLE
[driver/cc1depscanProtocol] For the spawned depscan daemon process close all file descriptors that are not explicitly inherited

### PR DESCRIPTION
**Pull request by akyrtzi**
This is using a darwin-specific extension, we need something different for other platforms.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: clang/tools/driver/cc1depscanProtocol.cpp [hunk 1]
```diff
@@ -169,11 +170,28 @@ Expected<ScanDaemon> ScanDaemon::launchDaemon(StringRef BasePath,
...
+  if (int EC = posix_spawnattr_setflags(&attrs, POSIX_SPAWN_CLOEXEC_DEFAULT))
+    return llvm::errorCodeToError(std::error_code(EC, std::generic_category()));
+#endif
+
   ::pid_t Pid;
-  int EC =
-      ::posix_spawn(&Pid, Args[0], /*file_actions=*/nullptr,
-                    /*attrp=*/nullptr, const_cast<char **>(LaunchArgs.data()),
-                    /*envp=*/nullptr);
+  int EC = ::posix_spawn(&Pid, Args[0], /*file_actions=*/nullptr, &attrs,
```
Comment (cachemeifyoucan) [inline clang/tools/driver/cc1depscanProtocol.cpp]: I wonder if calling with `attrs == nullptr` is the same as an empty attrs. From documentation, `nullptr` should inherit everything from the parent(?) while the empty set might reset the attribute? Not sure but maybe someone would know the detail.

**Review by cachemeifyoucan — COMMENTED** [review]
Just a concern inline. Otherwise LGTM as I don't know how to write a test for this.

File: clang/tools/driver/cc1depscanProtocol.cpp (hunk 1 above)
Comment (dexonsmith) [reply clang/tools/driver/cc1depscanProtocol.cpp]: From the `posix_spawnattr_init` manpage, it seems like the default / empty set is likely to match the parent:
```
     The posix_spawnattr_init() function initializes the object referenced by
     attr, to an empty set of spawn attributes for subsequent use in a call to
     posix_spawn(2) or posix_spawnp(2).

     A posix_spawnattr_t object is an opaque type that is used to collect set of
     spawn attributes specified by calls to posix_spawnattr_setflags(3),
     posix_spawnattr_setpgroup(3), posix_spawnattr_setsigdefault(3),
     posix_spawnattr_setsigmask(3), posix_spawnattr_setarchpref_np(3),
     posix_spawnattr_setbinpref_np(3), posix_spawnattr_setspecialport_np(3), or
     posix_spawnattr_setexceptionports_np(3) such that once the spawn operation
     is complete, the child process will have a different set of attributes than
     the parent.  The order of these operations is unspecified and may change;
     programs should not depend on their ordering.
```
"[...] such that once the spawn operation is complete, the child process will have a different set of attributes than the parent"

My reading is that this is fine as-is.

Comment (akyrtzi) [reply clang/tools/driver/cc1depscanProtocol.cpp]: My reading is same as @dexonsmith; from what I understand `posix_spawn` has default behavior which can be overridden using user specified spawn attributes, and whether you pass `NULL` or an empty set it has the same effect (no change in default behavior).

**Comment by akyrtzi** [comment]
Forgot pascal case for the new local variables, fixed now.

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by benlangmuir** [comment]
*Is* there a way to do this portably? The best approach I could think of is to loop over all possible file descriptors (you can use `getrlimit` to find the maximum fd) and either close them except for stdout/stdin/stderr in the child process, or fcntl them FD_CLOEXEC in the parent process.  But if the maximum fd is INT_MAX, that would suck.

**Comment by akyrtzi** [comment]
> _Is_ there a way to do this portably? The best approach I could think of is to loop over all possible file descriptors (you can use `getrlimit` to find the maximum fd) and either close them except for stdout/stdin/stderr in the child process, or fcntl them FD_CLOEXEC in the parent process. But if the maximum fd is INT_MAX, that would suck.

Alternatively for other platforms we could just handle the `llbuild` issue and close its file descriptor. It is found in the `LLBUILD_CONTROL_FD` environment variable.

**Pull request merged**